cmd/gateway-azure.go

```diff
@@ -639,14 +656,47 @@ func (a *azureObjects) CompleteMultipartUpload(bucket, object, uploadID string,
 	if meta == nil {
 		return objInfo, traceError(InvalidUploadID{uploadID})
 	}
+
+	resp, err := a.client.GetBlockList(bucket, object, storage.BlockListTypeUncommitted)
+	if err != nil {
+		return objInfo, azureToObjectError(traceError(err), bucket, object)
+	}
+
+	getBlocks := func(partNumber int, etag string) (blocks []storage.Block, err error) {
+		var idBytes []byte
+		for _, part := range resp.UncommittedBlocks {
+			if idBytes, err = base64.StdEncoding.DecodeString(part.Name); err != nil {
+				return nil, err
+			}
+
+			id := string(idBytes)
+			if strings.HasPrefix(id, strconv.Itoa(partNumber)+".") && strings.HasSuffix(id, "."+etag) {
+				blocks = append(blocks, storage.Block{
+					ID:     part.Name,
+					Status: storage.BlockStatusUncommitted,
+				})
+			}
+		}
+
+		if len(blocks) == 0 {
+			return nil, InvalidPart{}
+		}
+
+		return blocks, nil
+	}
```

Comment:
introducing azureEncodeBlockID and azureDecodeBlockID will make the above code more readable:
```
func azureEncodeBlockID(partID, subPartNumber int, md5Hex string) string {                              
        return base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%05d.%02d.%s", partID, subPartNumb\
er, md5Hex)))                                                                                           
}                                                                                                       
                                                                                                        
func azureDecodeBlockID(blockID string) (partID, subPartNumber int, md5Hex string, err error) {         
        b, err := base64.StdEncoding.DecodeString(blockID)                                              
        if err != nil {                                                                                 
                return                                                                                  
        }                                                                                               
        fmt.Sscanf(string(b), "%05d.%02d.%s", &partID, &subPartNumber, &md5Hex)                         
        return                                                                                          
}                                                                                                       

....
...

        getBlocks := func(partNumber int, etag string) (blocks []storage.Block, err error) {            
                var idBytes []byte                                                                      
                for _, part := range resp.UncommittedBlocks {                                           
                        respPartNumber, _, respEtag, err := azureDecodeBlockID(part.Name)               
                        if err != nil {                                                                 
                                return nil, InvalidPart{}                                               
                        }                                                                               
                        if partNumber == respPartNumber && etag == respEtag {                           
                                blocks = append(blocks, storage.Block{                                  
                                        ID:     part.Name,                                              
                                        Status: storage.BlockStatusUncommitted,                         
                                })                                                                      
                        }                                                                               
                }                                                                                       
                                                                                                        
                if len(blocks) == 0 {                                                                   
                        return nil, InvalidPart{}                                                       
                }                                                                                       
                                                                                                        
                return blocks, nil                                                                      
        }                                                                                               


```

Reply:
Done